packages/hash/frontend/src/components/RemoteBlock/RemoteBlock.tsx

```diff
@@ -16,54 +30,99 @@ type RemoteBlockProps = {
  */
 export const RemoteBlock: React.VFC<RemoteBlockProps> = ({
-  blockFunctions,
-  blockProperties,
+  blockMetadata,
   crossFrame,
   editableRef,
-  sourceUrl,
+  graphCallbacks,
+  graphProperties,
   onBlockLoaded,
+  sourceUrl,
 }) => {
-  const [loading, err, Component] = useRemoteBlock(
+  const [loading, err, blockSource] = useRemoteBlock(
     sourceUrl,
     crossFrame,
     onBlockLoaded,
   );
 
+  const wrapperRef = React.useRef<HTMLDivElement>(null);
+
+  const { graphService } = useGraphEmbedderService(wrapperRef, {
+    callbacks: graphCallbacks,
+    ...graphProperties,
+  });
+
+  React.useEffect(() => {
+    if (graphService) {
+      graphService.blockEntity({ data: graphProperties.blockEntity });
+    }
+  }, [graphProperties.blockEntity, graphService]);
+
+  React.useEffect(() => {
+    if (graphService) {
+      graphService.blockGraph({ data: graphProperties.blockGraph });
+    }
+  }, [graphProperties.blockGraph, graphService]);
+
+  React.useEffect(() => {
+    if (graphService) {
+      graphService.entityTypes({ data: graphProperties.entityTypes });
+    }
+  }, [graphProperties.entityTypes, graphService]);
+
+  React.useEffect(() => {
+    if (graphService) {
+      graphService.linkedAggregations({
+        data: graphProperties.linkedAggregations,
+      });
+    }
+  }, [graphProperties.linkedAggregations, graphService]);
+
   if (loading) {
     return <BlockLoadingIndicator />;
   }
 
-  if (!Component) {
+  if (!blockSource) {
     throw new Error("Could not load and parse block from URL");
   }
 
   if (err) {
     throw err;
   }
 
-  if (typeof Component === "string") {
-    /**
-     * This HTML block has no props available to it, unless loaded via FramedBlock.
-     * @todo do something about this. throw if not in an iframe?
-     *    or check for iframe status and assign props to window here, not FramedBlock?
-     */
-    return (
-      <HtmlBlock
-        html={Component}
-        blockFunctions={blockFunctions}
-        blockProperties={blockProperties}
-      />
-    );
-  }
+  const propsToInject: BlockGraphProperties<Record<string, any>> & {
+    editableRef: any;
+  } = {
+    editableRef,
+    graph: graphProperties,
+  };
+
+  // @todo remove this React default when we update all blocks to 0.2
+  const entryPoint = blockMetadata.blockType?.entryPoint ?? "react";
 
   return (
-    <Component
-      {...blockFunctions}
-      {...blockProperties}
-      editableRef={editableRef}
-    />
+    <div ref={wrapperRef}>
+      {graphService ? (
+        <BlockRenderer
+          customElement={
+            (entryPoint === "custom-element" ? blockSource : undefined) as any
+          }
+          html={
+            (entryPoint === "html"
+              ? { source: blockSource, url: sourceUrl }
+              : undefined) as any
+          }
+          properties={propsToInject}
+          ReactComponent={
+            (entryPoint === "react" ? blockSource : undefined) as any
+          }
```

Comment:
Good shout, I think I cast these to sort out later and then never did.

Addressed partially in https://github.com/hashintel/hash/pull/752/commits/d22af8b1c1f4e692cc9434e215d43221ce1aaaa1 in that there's runtime checking that `entryPoint` matches the expected `blockSource` – the compiler isn't narrowing it based on the checking so it could probably be improved